# STAT 220: Lab 2, Red Dragon Inn

In Dungeons and Dragons, the notation mdn denotes the process of rolling m dice, each with n sides. Here, rolling two six-sided dice is expressed as 2d6, and rolling three eight-sided dice is denoted by 3d8. You are tasked with creating a utility to assist players in the following two scenarios:

A. Calculating the probability of rolling m dice, each with n sides, and having the sum of the dice be greater than x. This calculation should only be implemented for m = 1 through 4 and n is at most 20.

B. Determining the probability of rolling m dice, each with n sides, and having at least one die show a value greater than or equal to x.

Your task is to develop two functions, one for each scenario, where users can input values for m , n, and x to compute the required probability. Ensure to document your code well, with comments that effectively guide users through each step of the computation process.

## Probability Functions

In [42]:
import itertools
import numpy as np

def ProbGreaterThanX(m, n, x):
  # Create a list of all possible rolls of m dice with n sides
  rolls = list(itertools.product(range(1, n+1), repeat=m))
  # Calculate the sum of each roll
  sums = np.array([sum(roll) for roll in rolls])
  # Calculate the proportion of rolls that have a sum greater than or equal to x
  return np.mean(sums >= x)

def ProbAtLeastOneGreaterThanX(m, n, x):
  # Create a list of all possible rolls of m dice with n sides
  rolls = [list(roll) for roll in itertools.product(range(1, n + 1), repeat=m)]
  success = []
  for roll in rolls:
        if max(roll) >= x:
            success.append(1)
        else:
            success.append(0)
  return np.mean(success)

print(ProbGreaterThanX(2, 6, 7))
print(ProbAtLeastOneGreaterThanX(2, 6, 3))

0.5833333333333334
0.8888888888888888


## Evaluate the performance
of your tool using the following scenario:
A player is in a challenging situation where they have a choice. They can choose between the following challenges:

• Roll 1d20 and get a result of 17 or more

• Roll 3d4 and get a result of 10 or more

• Roll 4d6 and have at least one dice show a 6


Use your tool to compute the probabilities and report the best choice for the player.

In [44]:
print("Challenge 1: ", ProbGreaterThanX(1, 20, 17))
print("Challenge 2: ", ProbGreaterThanX(3, 4, 10))
print("Challenge 3: ", ProbAtLeastOneGreaterThanX(4, 6, 6))

Challenge 1:  0.2
Challenge 2:  0.15625
Challenge 3:  0.5177469135802469


**Response:** Based on the results, I would recommend the player to choose the challenge of rolling 4d6 and having at least one die show a 6. This challenge has the highest probability of success at 51.77%. The first two challenges have probabilities of only 20% and 15.63%, respectively.